composer/datasets/utils.py

```diff
@@ -186,20 +186,22 @@ def __init__(
             self.stop_sequence = stop_sequence
             self.stop_sequence_ids = tokenizer.encode(stop_sequence, add_special_tokens=False)
 
-            # we look back for 2 more tokens than it takes to encode our stop sequence
+            # sentence piece tokenizers add a superflous underline token before string-initial \n
+            # that throws off our calculation of the stop sequence length
+            # so we remove any token ids that produce empty strings
+            self.stop_sequence_ids = [id for id in self.stop_sequence_ids if tokenizer.decode(id) != '']
+
+            # we look back for 1 more token than it takes to encode our stop sequence
             # because tokenizers suck, and a model might generate `['\n', '\n']` but our `sequence` is `['\n\n']`
             # and we don't want to mistakenly not stop a generation because our
             # (string) stop sequence was output in a different tokenization
 
-            # NOTE: there is a minor danger that this will end up looking back 2 tokens into the past, into the inputs to the model,
-            # and stopping generation immediately as a result. With only 2 extra tokens of lookback, this risk is minimized
-            self.stop_sequence_id_len = len(self.stop_sequence_ids) + 2
+            self.stop_sequence_id_len = len(self.stop_sequence_ids) + 1
```

Comment:
why did you change this to +1

Reply:
since the model generates 1 token before the first call to the stopping criteria, this ensures we will never stop too early for any of the stopping criteria we use in practice (since they're all multi-token).